File: app/models/sale.py
"""
sale.py — Modelo de Venta completa.
"""
from ..database.db import db
from ..utils import now_mx

class Sale(db.Model):
    __tablename__ = "sales"
    id = db.Column(db.Integer, primary_key=True)
    client_name = db.Column(db.String(150), default="")
    subtotal_amount = db.Column(db.Float, default=0.0)
    discount_amount = db.Column(db.Float, default=0.0)
    tax_amount = db.Column(db.Float, default=0.0)
    total_amount = db.Column(db.Float, nullable=False)
    payment_method = db.Column(db.String(50), default="efectivo")
    payment_status = db.Column(db.String(20), default="aprobado")
    created_at = db.Column(db.DateTime, default=now_mx)
    customer_id = db.Column(db.Integer, db.ForeignKey("customers.id"), nullable=True)
    branch_id   = db.Column(db.Integer, db.ForeignKey("branches.id"), nullable=True)
    items = db.relationship("SaleItem", backref="sale", lazy=True, cascade="all, delete-orphan")

    def to_dict(self):
        return {"id": self.id, "client_name": self.client_name or "", "customer_id": self.customer_id,
                "subtotal_amount": self.subtotal_amount or 0.0,
                "discount_amount": self.discount_amount or 0.0,
                "tax_amount": self.tax_amount or 0.0,
                "total_amount": self.total_amount,
                "payment_method": self.payment_method or "efectivo",
                "payment_status": self.payment_status or "aprobado",
                "branch_id": self.branch_id,
                "branch_name": self.branch.name if self.branch else None,
                "created_at": self.created_at.isoformat(),
                "items": [i.to_dict() for i in self.items]}

File: flutter_seller/lib/models/sale.dart
import 'sale_item.dart';

class Sale {
  final int id;
  final String clientName;
  final double totalAmount;
  final DateTime createdAt;
  final List<SaleItem> items;

  const Sale({
    required this.id,
    required this.clientName,
    required this.totalAmount,
    required this.createdAt,
    required this.items,
  });

  factory Sale.fromJson(Map<String, dynamic> json) {
    return Sale(
      id: json['id'] as int,
      clientName: (json['client_name'] as String?) ?? '',
      totalAmount: (json['total_amount'] as num).toDouble(),
      createdAt: DateTime.parse(json['created_at'] as String),
      items: (json['items'] as List<dynamic>?)
              ?.map((e) => SaleItem.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }
}
